llmfoundry/callbacks/hf_checkpointer.py

```diff
@@ -0,0 +1,165 @@
+    """Save a huggingface formatted checkpoint during training.
+
+    Args:
+        save_folder (str): Top level folder to save checkpoints to (can be a URI). It is likely that
+            this would be the same as your save_folder.
+        save_interval: Union[str, int, Time]: The interval describing how often checkpoints should be
+            saved. If an integer, it will be assumed to be in :attr:`.TimeUnit.EPOCH`.
+            Otherwise, the unit must be either :attr:`.TimeUnit.EPOCH`, :attr:`.TimeUnit.BATCH`,
+            :attr:`.TimeUnit.TOKEN`, or :attr:`.TimeUnit.SAMPLE`.
+        huggingface_folder_name (str): Folder to save each checkpoint under (can be a format string). Default is ``ba{batch}``.
+        precision: The precision to save the model in. Default is ``float32``. Options are ``bfloat16``, ``float16``, or ``float32``.
+        overwrite (bool): Whether to overwrite previous checkpoints.
+    """
+
+    def __init__(
+        self,
+        save_folder: str,
+        save_interval: Union[str, int, Time],
+        huggingface_folder_name: str = 'ba{batch}',
+        precision: str = 'fp32',
+        overwrite: bool = False,
+    ):
+        self.backend, self.bucket_name, self.save_dir_format_str = parse_uri(
+            save_folder)
+        self.overwrite = overwrite
+        self.precision = precision
+        self.dtype = {
+            'float32': torch.float32,
+            'float16': torch.float16,
+            'bfloat16': torch.bfloat16,
+        }[precision]
+        self.huggingface_folder_name_fstr = os.path.join(
+            'huggingface', huggingface_folder_name)
+        self.check_interval = create_interval_scheduler(
+            save_interval, include_end_of_training=True)
+        self.upload_to_object_store = (self.backend != '')
+        if self.upload_to_object_store:
+            self.remote_ud = RemoteUploaderDownloader(
+                bucket_uri=f'{self.backend}://{self.bucket_name}',
+                num_concurrent_uploads=4)
+        else:
+            self.remote_ud = None
+
+        self.last_checkpoint_batch: Optional[Time] = None
+
+    def run_event(self, event: Event, state: State, logger: Logger) -> None:
+        # The interval scheduler handles only returning True for the appropriate events
+        if state.get_elapsed_duration() is not None and self.check_interval(
+                state,
+                event) and self.last_checkpoint_batch != state.timestamp.batch:
+            self._save_checkpoint(state, logger)
+        elif event == Event.INIT:
+            if not isinstance(state.model, HuggingFaceModel):
+                raise ValueError(
+                    f'`HuggingFaceCheckpointer` is only compatible with `HuggingFaceModel`s. '
+                    + f'Got {type(state.model)} instead.')
+            if self.upload_to_object_store and self.remote_ud is not None:
+                self.remote_ud.init(state, logger)
+                state.callbacks.append(self.remote_ud)
```

Comment:
is it appended to callbacks and not loggers because the engine is just given callbacks, which has all the loggers?

Reply:
again, copied from the monolithic saver haha

Reply:
haha forgetting my own code lol